proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,265 @@
+   
+   num_partitions: u64, // little-endian unsigned 64-bit integer
+   
+   parent_blockhash: Hash, // byte-array of length 32
+
+   // total points calculated for the current epoch, where points equals the sum
+   // of delegated stake * credits observed for all delegations
+   total_points: u128, // little-endian unsigned 128-bit integer
+
+   // total rewards for the current epoch, in lamports
+   total_rewards: u64, // little-endian unsigned 64-bit integer
+
+   // distributed rewards for the current epoch, in lamports
+   distributed_rewards: u64, // little-endian unsigned 64-bit integer
+}
+```
+
+The `EpochRewards` sysvar is repopulated at the start of the first block of the
+epoch (before any transactions are processed), as both the total epoch rewards
+and vote account rewards become available at this time. The
+`distributed_rewards` field is updated per reward distribution for each block in
+the reward distribution phase. The `active` field is set to false after the last
+partition is distributed (and `total_rewards == distributed_rewards`), ie.
+before processing transactions in block at height `X + M`.
+
+#### Booting from Snapshot
+
+When booting from a snapshot, a node must check the EpochRewards sysvar account
+to determine whether the distribution phase is active. If so, the node must
+rerun the rewards partitioning using the `EpochRewards::num_partitions` and
+`EpochRewards::parent_blockhash` sysvar fields and determining the upcoming
+partitions by comparing its current block height to
+`EpochRewards::distribution_starting_block_height`. Then the runtime must
+recalculate the remaining rewards using the `EpochRewards::total_points` and
+`EpochRewards::total_rewards` sysvar fields, as well as the `EpochStakes` in the
+snapshot. The recalculated rewards can be confirmed by comparing a sum of the
+rewards remaining (those partitions expected to not yet have been distributed)
+with the difference between the `EpochRewards::total_rewards` and
+`EpochRewards::distributed_rewards` fields. Partitions for blocks prior to the
+current block height can be discarded.
+
+### Restrict Stake Account Mutation
+
+The stake accounts need to be preserved in credit-only state while distributions
+are in progress in order to prevent any changes interfering with reward
+distribution.
+
+Therefore, the Stake Program needs access to a syscall which reports whether the
+distribution phase is active. This new syscall `sol_get_epoch_rewards_sysvar`
+should return the values of the `EpochRewards` sysvar. All Stake Program
+instructions that mutate stake data or debit stake balances must be disabled
+when `EpochRewards::active` is true.
+
+Any transaction that attempts to invoke such an instruction will fail with this
+new error code:
+
+```
+StakeError {
+   EpochRewardsDistributionActive = 15,
+}
+```
+
+Other users can access the `active` field from the
+`sol_get_epoch_rewards_sysvar` syscall to determine the distribution-phase
```

Comment:
Is this syscall going to be specified by another SIMD?

Reply:
we probably need a generic "sysvar getter syscall" simd to cover stuff like this

Reply:
@ripatel-fd , what do you think of Trent's idea here? There are a lot of these sysvar-getter syscalls already, and it would be nice not to have to define new ones beyond the sysvar layout in SIMDs like this one.